allensdk/brain_observatory/ecephys/current_source_density/_current_source_density.py

```diff
@@ -189,102 +135,57 @@ def accumulate_lfp_data(timestamps, lfp_raw, lfp_channels, trial_windows, extrac
+    lfp_stack: numpy.ndarray
+        Stack of LFP traces surrounding presentation of a common stimulus.
+        Dimensions are trials X channels X samples.
+    spacing : float
+        Distance between channels, in millimeters. This spacing may be
+        physical distances between channels or a virtual distance if channels
+        have been interpolated to new virtual positions.
 
     Returns
     -------
-    csd : numpy.ndarray
-        current source density. Dimensions are channels X samples
-
+    Tuple[csd, csd_channels]:
+        csd : numpy.ndarray
+            Current source density. Dimensions are channels X samples.
+        csd_channels: numpy.ndarray
+            Array of channel indices for CSD.
     '''
 
-    accumulated_lfp = accumulated_lfp - np.mean(accumulated_lfp, axis=2)[:, :, None]
-    lfp = np.mean(accumulated_lfp, axis=0)
+    # mean_lfp = np.zeros((np.shape(lfp_stack)[1] + 2,
+    #                      np.shape(lfp_stack)[2]))
 
-    lfp, channels = resample_to_regular(lfp, channels, missing_channels)
+    # mean_lfp[1:-1,:] = np.nanmean(lfp_stack, axis=0)
+    # mean_lfp[0,:] = mean_lfp[1,:]
+    # mean_lfp[-1,:] = mean_lfp[-2,:]
 
-    lfp_smooth=np.zeros([np.shape(lfp)[0] - 2, np.shape(lfp)[1]])
-    csd = np.zeros([np.shape(lfp_smooth)[0] - 2, np.shape(lfp_smooth)[1]])
-    
-    for t in range(np.shape(lfp)[1]):  # time
-        for d in range(1, np.shape(lfp)[0] - 1):  # depth
-            lfp_smooth[d - 1, t] = 0.25 * ( lfp[d - 1, t] + 2 * lfp[d, t] + lfp[d + 1, t] )
-    
-    for t in range(np.shape(lfp_smooth)[1]): # time
-        for d in np.arange(1, np.shape(lfp_smooth)[0] - 1):
-            csd[d - 1, t]=(1 / spacing ** 2) * ( lfp_smooth[d + 1, t] - 2 * lfp_smooth[d, t] + lfp_smooth[d - 1, t] )    
-    
-    return csd, channels
+    mean_lfp = np.nanmean(lfp_stack, axis=0)
+    mean_lfp = np.pad(mean_lfp, pad_width=((1, 1), (0, 0)), mode='edge')
 
+    csd = np.zeros((lfp_stack.shape[1], lfp_stack.shape[2]))
 
-def resample_to_regular(lfp, channels=None, missing_channels=None):
-    ''' Linearly interpolates in order to fill in channel gaps in a 2D LFP array.
-    
-    Parameters
-    ----------
-    lfp : numpy.ndarray
-        2D, with dimensions channel X sample
-    channels : numpy.ndarray, optional
-        Indices of channels whose LFP data can be used directly 
-    missing_channels : dict, optional
-        get_missing_channels output
-    
-    Returns
-    -------
-    new_lfp : numpy.ndarray
-        Same dimensions as input LFP, but missing channel data has been interpolated.
-    total_channels : numpy.ndarray
-        Array of channel indices, real and interpolated, that correspond to rows of the interpolated LFP image.
+    for t in range(np.shape(mean_lfp)[1]):  # time
+        for d in np.arange(1, np.shape(mean_lfp)[0] - 1):  # space
 
-    '''
+            csd[d - 1, t] = (1 / (spacing ** 2)) * (mean_lfp[d + 1, t]
```

Comment:
This should be vectorized via matrix multiplication. There are several methods of computing csd. Thus, it is better to move this particular method into a separate function.

Reply:
Vectorized, but let me know if there are even more efficient ways I can do this.